core/src/main/java/io/questdb/griffin/engine/orderby/LongSortedLightRecordCursorFactory.java

```diff
@@ -0,0 +1,130 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.orderby;
+
+import io.questdb.cairo.AbstractRecordCursorFactory;
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.ListColumnFilter;
+import io.questdb.cairo.sql.RecordCursor;
+import io.questdb.cairo.sql.RecordCursorFactory;
+import io.questdb.cairo.sql.RecordMetadata;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+
+/**
+ * Radix sort-based factory for ORDER BY over a single int/ipv4/long/timestamp/date column.
+ * Memory overhead is similar to {@link SortedRecordCursorFactory}:
+ * each row takes 32 bytes vs 36 bytes in the red-black tree-based factory.
+ */
+public class LongSortedLightRecordCursorFactory extends AbstractRecordCursorFactory {
+    private final RecordCursorFactory base;
+    private final LongSortedLightRecordCursor cursor;
+    private final ListColumnFilter sortColumnFilter;
+
+    public LongSortedLightRecordCursorFactory(
+            CairoConfiguration configuration,
+            RecordMetadata metadata,
+            RecordCursorFactory base,
+            ListColumnFilter sortColumnFilter
+    ) {
+        super(metadata);
+        this.base = base;
+        final int columnIndex = sortColumnFilter.getColumnIndexFactored(0);
+        this.cursor = new LongSortedLightRecordCursor(
+                configuration,
+                columnIndex,
+                metadata.getColumnType(columnIndex),
+                sortColumnFilter.getColumnIndex(0) > 0
+        );
+        this.sortColumnFilter = sortColumnFilter;
+    }
+
+    public static void addSortKeys(PlanSink sink, ListColumnFilter filter) {
+        sink.attr("keys").val('[');
+        filter.toPlan(sink);
+        sink.val(']');
+    }
+
+    public static boolean isSupportedColumnType(int columnType) {
+        return columnType == ColumnType.INT
+                || columnType == ColumnType.IPv4
+                || columnType == ColumnType.LONG
+                || columnType == ColumnType.TIMESTAMP
+                || columnType == ColumnType.DATE;
+    }
+
+    @Override
+    public RecordCursorFactory getBaseFactory() {
+        return base;
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor baseCursor = base.getCursor(executionContext);
+        try {
+            cursor.of(baseCursor, executionContext);
+            return cursor;
+        } catch (Throwable ex) {
+            baseCursor.close();
+            cursor.close();
```

Comment:
this is ripe for duble-free. We must enforce predictable behavior out of "of" method. As in, it must be responsible for closing the "baseCursor". The problem exists in other classes that call "of"

Reply:
Good spot. Fixed it in this and other places in 383831d31507a93c7f83849e85f0e82dcd46a646